app/notification_push.go

```diff
@@ -322,9 +338,14 @@ func (hub *PushNotificationsHub) stop() {
 			notificationType: notificationTypeDummy,
 		}
 	}
-	hub.stopChan <- struct{}{}
-	close(hub.notificationsChan)
+	close(hub.stopChan)
+	// We need to wait for the outer for loop to exit first.
+	// We cannot just send struct{}{} to stopChan because there are
+	// other listeners to the channel. And sending just once
+	// will cause a race.
 	hub.wg.Wait()
+	// And then we wait for the semaphore to finish.
+	hub.semaWg.Wait()
```

Comment:
Now that we rely on a wait group do we still need to send the dummy notification on stop? 
I am wondering if adding a case for `<-hub.stopChan` also in the inner notification handler goroutine would be sufficient.

Reply:
The dummy notifications are meant to drain any pending notifications in the channel buffer. Adding a case for `<-hub.stopChan` will just clear all in-flight goroutines right? Will the remaining notifications still in the channel buffer be handled?

Reply:
Not really familiar with this part of the codebase so please forgive my confusion.

I am missing where the drain happens. I can see we are filling `hub.notificationsChan` and as a consequence starting goroutines  which I find slightly odd since what we really want is to stop. 

As far as I can see, as soon as we close `hub.stopChan`, no more notifications will be received so I am questioning whether that would be enough to shut down instead of sending these dummy messages.

Reply:
> As far as I can see, as soon as we close hub.stopChan, no more notifications will be received 

That is correct. But closing the `stopChan` and waiting on `semaWg` does not guarantee that the pending notifications in the channel buffer will be handled. Because closing it will immediately exit from the method, so there's nothing consuming from the channel any more. Right? Unless I am missing something? And we will only be waiting for pending goroutines which are already in-flight to be over.

Hence, my idea was to push the buffer-length number of dummy notifications _after_ shutting down the HTTP server, so that whatever notifications are there in the buffer gets pushed out and handled. And _then_ exit the listener and wait for semaphore to return.

Does that make sense?

It is a bit tricky and even I am not 100% sure if I have covered all the cases. :P So it might well be that there is something I have missed.

Reply:
> That is correct. But closing the `stopChan` and waiting on `semaWg` does not guarantee that the pending notifications in the channel buffer will be handled. Because closing it will immediately exit from the method, so there's nothing consuming from the channel any more. Right? Unless I am missing something? And we will only be waiting for pending goroutines which are already in-flight to be over.

Alright so I guess the question is: do we care about handling what's left in the buffer when shutting down? Worst thing that can happen is it gets full but then we are already making sure the writes are non-blocking.

Reply:
The thing is that this channel buffer is configurable by the user. And the default is 1000. In a cloud environment where pods are frequently restarted, my fear is that considerable number of push notifications can get lost.

Reply:
I see. Obviously since we are stopping the handler some sort of loss is unavoidable. What still leaves me a bit confused is the need to spawn the dummy goroutines. If the goal is handling whatever notification is left wouldn't it be simpler if we did it by looping the length of the buffered channel as soon as we get the signal that `hub.stopChan` was closed in `start()`, just before returning and exiting the main loop?

Reply:
Ahh I see. Yeah not bad. We did something similar in the websocket handler in the load test code right?

Reply:
I believe that was a bit different since we didn't really care about processing the backlog of events. Also in that case we were able to guarantee no more sends to the event channel which could be eventually closed. Here it's trickier.